Return string

mmmries opened the pull request:
Hey @JonnerLynx I made some progress on this after our pairing session this afternoon. In this PR I've made a few improvements like having a single static context that gets initialized when our NIF is loaded into the VM, some basic argument checking, and a benchmark so we can keep an eye on the performance of our overall solution.

I'm seeing a strange behavior though, where calling `enif_inspect_binary` creates the `ErlNifBinary` with the correct `size`, but checking `strlen(data)` shows an extra byte on the end of the string? This is resulting in mpdecimal failing to parse the strings we pass to it and producing a `NaN` result.

My guess is that we should be using the `size` attribute of the `ErlNifBinary` when parsing the string, rather than relying on null-termination? But I'm not sure how we can fix the string? Or maybe we should just be using a different function for parsing the strings?

Here are the benchmark results (with the printf's removed)

```
Operating System: macOS
CPU Information: Intel(R) Core(TM) i9-8950HK CPU @ 2.90GHz
Number of Available Cores: 12
Available memory: 16 GB
Elixir 1.11.2
Erlang 23.1.1

Benchmark suite executing with the following configuration:
warmup: 2 s
time: 5 s
memory time: 0 ns
parallel: 1
inputs: none specified
Estimated total run time: 21 s

Benchmarking 2 ^ 2...
Benchmarking 2.0 ^ 2.0...
Benchmarking 3.14159 ^ 1.58732...

Name                        ips        average  deviation         median         99th %
2 ^ 2                  958.78 K        1.04 μs   ±139.90%           1 μs           2 μs
2.0 ^ 2.0              884.66 K        1.13 μs  ±1317.34%           1 μs           2 μs
3.14159 ^ 1.58732       12.32 K       81.17 μs    ±20.75%          78 μs         161 μs

Comparison: 
2 ^ 2                  958.78 K
2.0 ^ 2.0              884.66 K - 1.08x slower +0.0874 μs
3.14159 ^ 1.58732       12.32 K - 77.83x slower +80.13 μs
```

JonnerLynx commented:
Hey @mmmries, this is great progress! The anomaly in reported size due to the fact that cstrings rely on NULL termination, while Erlang binaries do not. This means that when interpreting an Erlang binary as a cstring, it's possible to encounter a fair number of garbage bytes before a NULL character happens to appear in the sequence. As you would expect, this isn't deterministic between separate invocations of the `power()` function, as Erlang makes no guarantees about the value of the bytes in memory that come after the designated `size` number of bytes.

This also explains why NaN is a fairly common result, as the mpdecimal library does expect null terminated strings. This certainly has the possibility of breaking benchmarks, so we'll need to `strcpy` the correct number of bytes into a cstring for consumption by mpdecimal.

```
iex(1)> MPDecimal.Nif.power("2.0", "2.0")
base  (as binary): <<"2.0">>
power (as binary): <<"2.0">>
base  (as cstring): 2.0
  - binary size: 3
  - cstring strlen: 3
power (as cstring): 2.0
  - binary size: 3
  - cstring strlen: 3
base encoding:  0x32 0x2e 0x30 0x00
power encoding: 0x32 0x2e 0x30 0x00
"4.00"
iex(2)> MPDecimal.Nif.power("2.0", "3.0")
base  (as binary): <<"2.0">>
power (as binary): <<"3.0">>
base  (as cstring): 2.0��
  - binary size: 3
  - cstring strlen: 6
power (as cstring): 3.0��
  - binary size: 3
  - cstring strlen: 6
base encoding:  0x32 0x2e 0x30 0xfc 0xef 0x7f 0x00
power encoding: 0x33 0x2e 0x30 0xfc 0xef 0x7f 0x00
"NaN"
```

JonnerLynx commented:
One other note that explains the accumulating indentation that we were seeing with printf. It appears that Elixir's stdout parser expects carriage return characters in addition to newlines. So "\n" simply needs to be converted to "\r\n". Then the output appears as expected. Happy hacking!

mmmries commented on c_src/mpdecimal_nif.c:
```diff
@@ -10,16 +10,102 @@
+	// 
+	// TODO: Check whether the source of this data (Elixir Decimal library)
+	// guarantees this property.
+	char* base_cstring = (char*) malloc((base.size + 1)*sizeof(char));
```
Do you think we should use [enif_alloc](https://erlang.org/doc/man/erl_nif.html#enif_alloc) here? I don't have any specific reason to think that we do, but I noticed that they defined that function, so I thought there might be a reason they prefer that to the built-in allocation functions?

JonnerLynx replied:
Yes, good catch. I think it's a good idea to use the `enif_` flavor of any c standard library function that is available.

mmmries commented:
Your changes look awesome @JonnerLynx. I concur with your findings that we aren't getting conventional null-terminated strings for our binary terms. I pulled down your changes and tried them out, but I'm still seeing occasional `NaN` results. Here's an example iex session.

```
iex(1)> MPDecimal.Nif.power("3.14159", "1.58732")
base  (as binary): <<"3.14159">>
power (as binary): <<"1.58732">>
base  (as cstring): 3.14159
  - binary size: 7
  - cstring strlen: 7
power (as cstring): 1.58732
  - binary size: 7
  - cstring strlen: 7
base encoding:  0x33 0x2e 0x31 0x34 0x31 0x35 0x39 0x00 
power encoding: 0x31 0x2e 0x35 0x38 0x37 0x33 0x32 0x00 
"6.153686306332765322891030633"
iex(2)> MPDecimal.Nif.power("3.14159", "1.58732")
base  (as binary): <<"3.14159">>
power (as binary): <<"1.58732">>
base  (as cstring): 3.14159
  - binary size: 7
  - cstring strlen: 7
power (as cstring): 1.58732
  - binary size: 7
  - cstring strlen: 7
base encoding:  0x33 0x2e 0x31 0x34 0x31 0x35 0x39 0x00 
power encoding: 0x31 0x2e 0x35 0x38 0x37 0x33 0x32 0x00 
"NaN"
```

This doesn't seem to be related to the null-terminated strings since that change is in place and appears to be working, so I added a few commits that make it easy to turn debugging output on/off with a single `#define` and I added a few debug statements that print out the `a` and `b` mpd values after we load them from a string. I can see that in some cases these are being reported as a `NaN` even though the c-string looks to be properly formatted 🤔 

```
iex(12)> MPDecimal.Nif.power("3.14159", "1.58732")
base  (as binary): <<"3.14159">>
power (as binary): <<"1.58732">>
base  (as cstring): 3.14159
  - binary size: 7
  - cstring strlen: 7
power (as cstring): 1.58732
  - binary size: 7
  - cstring strlen: 7
base encoding:  0x33 0x2e 0x31 0x34 0x31 0x35 0x39 0x00 
power encoding: 0x31 0x2e 0x35 0x38 0x37 0x33 0x32 0x00 
mpd base: 3.14159
mpd power: NaN
mpd result: NaN
```

JonnerLynx commented:
It looks like we have some additional debugging to do on the string inputs to mpdecimal. What's odd to me is that with the exact same string inputs, we sometimes see a successful computation and we sometimes see a NaN result.

mmmries commented:
I added one more commit which pulls in the elixir decimal library as a dependency and it provides a top-level function that handles casting to strings, and then converting the result back to a `%Decimal{}`. I also added a benchmark to compare the performance of our nif to other decimal operations like addition and multiplication. You can run any of these benchmarks with just a basic `mix run bench/decimal_comparison.exs` invocation.

The good news is that it looks like we are only ~3x slower than the simple decimal operations provided by the library, but the bad news is that converting to/from strings on the elixir side adds ~1.8µs to our execution time.

```
Operating System: macOS
CPU Information: Intel(R) Core(TM) i9-8950HK CPU @ 2.90GHz
Number of Available Cores: 12
Available memory: 16 GB
Elixir 1.11.2
Erlang 23.1.1

Benchmark suite executing with the following configuration:
warmup: 1 s
time: 2 s
memory time: 0 ns
parallel: 1
inputs: none specified
Estimated total run time: 9 s

Benchmarking 2.0 * 2.0...
Benchmarking 2.0 + 2.0...
Benchmarking 2.0 ^ 2.0...

Name                ips        average  deviation         median         99th %
2.0 * 2.0        1.31 M      762.96 ns  ±1674.77%        1000 ns        1000 ns
2.0 + 2.0        1.23 M      813.79 ns  ±2373.78%        1000 ns        1000 ns
2.0 ^ 2.0        0.35 M     2840.66 ns   ±292.46%        3000 ns        5000 ns

Comparison: 
2.0 * 2.0        1.31 M
2.0 + 2.0        1.23 M - 1.07x slower +50.84 ns
2.0 ^ 2.0        0.35 M - 3.72x slower +2077.70 ns
```

mmmries commented:
Thanks for all your help on this. I'm feeling a bit stumped on a good next step so I'm going to try to keep my hands off this code for the rest of the weekend and see if any new ideas come up from "background" mental tasks.

JonnerLynx commented on lib/mpdecimal.ex:
```diff
@@ -1,2 +1,10 @@
+
+  def power(%Decimal{} = base, %Decimal{} = power) do
+    base = Decimal.to_string(base, :xsd)
+    power = Decimal.to_string(power, :xsd)
```
I think it makes the most sense to use the :xsd format, as it seem to be the most portable format available and only allows for the following characters: `+`, `-`, `.`, and `0-9`.

mmmries commented:
I just pulled down your changes and re-ran the benchmarks. Here are the numbers I got:

```
Operating System: macOS
CPU Information: Intel(R) Core(TM) i9-8950HK CPU @ 2.90GHz
Number of Available Cores: 12
Available memory: 16 GB
Elixir 1.11.2
Erlang 23.1.1

Benchmark suite executing with the following configuration:
warmup: 1 s
time: 2 s
memory time: 0 ns
parallel: 1
inputs: none specified
Estimated total run time: 15 s

Benchmarking 2 ^ 2...
Benchmarking 2.0 ^ 2.0...
Benchmarking 3.14159 ^ 1.58732...
Benchmarking 3.14159 ^ 1234.58732...
Benchmarking 9.0 ^ 0.5...

Name                           ips        average  deviation         median         99th %
2 ^ 2                     813.62 K        1.23 μs    ±88.22%           1 μs           2 μs
2.0 ^ 2.0                 763.24 K        1.31 μs    ±90.80%           1 μs           2 μs
9.0 ^ 0.5                  12.91 K       77.47 μs    ±17.87%          74 μs         154 μs
3.14159 ^ 1.58732          12.40 K       80.65 μs    ±16.66%          78 μs         157 μs
3.14159 ^ 1234.58732       11.28 K       88.66 μs    ±16.80%          85 μs      172.39 μs

Comparison: 
2 ^ 2                     813.62 K
2.0 ^ 2.0                 763.24 K - 1.07x slower +0.0811 μs
9.0 ^ 0.5                  12.91 K - 63.03x slower +76.24 μs
3.14159 ^ 1.58732          12.40 K - 65.62x slower +79.42 μs
3.14159 ^ 1234.58732       11.28 K - 72.13x slower +87.43 μs
riespro:mpdecimal$ mix run bench/decimal_comparison.exs 
make: Nothing to be done for `all'.
Operating System: macOS
CPU Information: Intel(R) Core(TM) i9-8950HK CPU @ 2.90GHz
Number of Available Cores: 12
Available memory: 16 GB
Elixir 1.11.2
Erlang 23.1.1

Benchmark suite executing with the following configuration:
warmup: 1 s
time: 2 s
memory time: 0 ns
parallel: 1
inputs: none specified
Estimated total run time: 9 s

Benchmarking 2.0 * 2.0...
Benchmarking 2.0 + 2.0...
Benchmarking 2.0 ^ 2.0...

Name                ips        average  deviation         median         99th %
2.0 * 2.0        1.33 M      751.66 ns  ±1763.07%        1000 ns        1000 ns
2.0 + 2.0        1.29 M      777.96 ns  ±2208.53%        1000 ns        1000 ns
2.0 ^ 2.0        0.34 M     2962.92 ns   ±356.40%        3000 ns        5000 ns

Comparison: 
2.0 * 2.0        1.33 M
2.0 + 2.0        1.29 M - 1.03x slower +26.30 ns
2.0 ^ 2.0        0.34 M - 3.94x slower +2211.26 ns
```

JonnerLynx commented:
New benchmark results. Note that comparing `Decimal.sqrt` to `MPDecimal.power` is not a fair comparison, as `MPDecimal.power` is the more general case that handles any exponent. (`MPDecimal.sqrt` tends to beat `Decimal.sqrt` by a healthy margin, but this is not shown here.)

Note that the runtime of `MPDecimal.power` is pretty consistently 70-80 μs, regardless of how trivial or complex the base and exponent are. This is presumably because the library is configured for 28 digits of precision and will always take a fixed number of passes in its internal algorithm to achieve that precision.

```Operating System: macOS
CPU Information: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
Number of Available Cores: 16
Available memory: 16 GB
Elixir 1.11.1
Erlang 23.3.1

Benchmark suite executing with the following configuration:
warmup: 1 s
time: 2 s
memory time: 0 ns
parallel: 1
inputs: none specified
Estimated total run time: 15 s

Benchmarking Decimal.sqrt("1234567890.1234567890123456")...
Benchmarking Decimal.sqrt("4.0")...
Benchmarking MPDecimal.power("1234567890.1234567890123456","0.1234567890123456789012345")...
Benchmarking MPDecimal.power("1234567890.1234567890123456","0.5")...
Benchmarking MPDecimal.power("4.0","0.5")...

Name                                                                                   ips        average  deviation         median         99th %
Decimal.sqrt("4.0")                                                               314.92 K        3.18 μs   ±318.29%        2.98 μs       13.98 μs
Decimal.sqrt("1234567890.1234567890123456")                                       214.39 K        4.66 μs   ±144.78%        3.98 μs       21.98 μs
MPDecimal.power("1234567890.1234567890123456","0.1234567890123456789012345")       13.06 K       76.59 μs    ±24.83%       71.98 μs      173.98 μs
MPDecimal.power("4.0","0.5")                                                       12.92 K       77.43 μs    ±28.53%       71.98 μs      175.35 μs
MPDecimal.power("1234567890.1234567890123456","0.5")                               12.07 K       82.84 μs    ±25.23%       75.98 μs      179.98 μs

Comparison:
Decimal.sqrt("4.0")                                                               314.92 K
Decimal.sqrt("1234567890.1234567890123456")                                       214.39 K - 1.47x slower +1.49 μs
MPDecimal.power("1234567890.1234567890123456","0.1234567890123456789012345")       13.06 K - 24.12x slower +73.42 μs
MPDecimal.power("4.0","0.5")                                                       12.92 K - 24.38x slower +74.25 μs
MPDecimal.power("1234567890.1234567890123456","0.5")                               12.07 K - 26.09x slower +79.66 μs```

JonnerLynx commented:
Through other tests not explicitly shown here, I was able to discover that the stringification overhead (Decimal -> string -> mpd -> string -> Decimal) is fixed at about 2 μs.

mmmries commented:
This looks awesome! Let's merge this as-is and then we can have one more PR to add some documentation and maybe write tests around known edge-cases before we release this and pull into `peg`?
